### PR TITLE
GEOMESA-1627 Expansion of Spark + SparkSQL docs

### PR DESCRIPTION
**Pull request by mdzimmerman**
With significant contributions from @Afterglow375 and @elahrvivaz.

* Documents geomesa-spark-core, geomesa-spark-sql
* Describes all current ``SpatialRDDProvider`` s.
* Lists all implemented UDTs and UDFs.
* Fixes aliasing of UDFs.
* Also contains some small fixes to Accumulo and Kafka tutorials.

Signed-off-by: Matthew Zimmerman <matt.zimmerman@ccri.com>

### CODE REVIEW

**Comment by mdzimmerman** [comment]
Replaces #1307.

**Comment by mdzimmerman** [comment]
Please add to the 1.3.0 milestone. Thanks!

**Comment by mdzimmerman** [comment]
Built version here: http://www.geomesa.org/documentation/fcr_spark_docs/index.html

File: docs/user/spark/core.rst [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    val conf = new SparkConf().setMaster("local[*]").setAppName("testSpark")
+    conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+    conf.set("spark.kryo.registrator", classOf[GeoMesaSparkKryoRegistrator].getName)
+    val sc = SparkContext.getOrCreate(conf)
+
+``geomesa-spark-core`` also provides an API for accessing geospatial data
+in Spark, by defining an interface called ``SpatialRDDProvider``. Different
+implementations of this interface connect to GeoMesa Accumulo, generic
+GeoTools-based ``DataStore``\ s, or data files in formats readable by the GeoMesa
+converter library. ``GeoMesaSpark`` loads an ``SpatialRDDProvider``
```
Comment (aheyne) [inline docs/user/spark/core.rst]: [picky] /an/a/

File: docs/user/spark/core.rst [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      // version, etc.
+    </dependency>
+
+``ConverterSpatialRDDProvider`` reads features from one or more data files in formats
+readable by the :doc:`/user/convert` library, including delimited and fixed-width text,
+Avro, JSON, and XML files. It takes the following configuration parameters:
+
+ * ``geomesa.converter`` - the converter defintion as a Typesafe Config string
+ * ``geomesa.converter.inputs`` - input file paths, comma-delimited
+ * ``geomesa.sft`` - the ``SimpleFeaturetype``, as a spec string, configuration string, or environment lookup name
```
Comment (aheyne) [inline docs/user/spark/core.rst]: /SimpleFeaturetype/SimpleFeatureType/

File: docs/user/spark/sparksql_functions.rst [hunk 3]
```diff
@@ -0,0 +1,805 @@
...
+.. _st_asGeoJSON:
+
+st_asGeoJSON
+^^^^^^^^^^^^
+
+::
+
+     String st_asGeoJSON(Geometry geom)
+
+Returns ``Geometry`` *geom* in
```
Comment (aheyne) [inline docs/user/spark/sparksql_functions.rst]: Looks like something is missing here.

File: docs/user/spark/sparksql_functions.rst [hunk 4]
```diff
@@ -0,0 +1,805 @@
...
+.. _st_relateBool:
+
+st_relateBool
+^^^^^^^^^^^^^
+
+::
+
+    Boolean st_relateBool(Geometry a, Geometry b, String mask)
+
+Returns true if the `DE-9IM`_ interaction matrix mask *mask* matches the interaction matrix pattern obtained from ``st_relate(a, b)``.
```
Comment (aheyne) [inline docs/user/spark/sparksql_functions.rst]: double mask

File: docs/user/spark/architecture.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
+Architecture
+------------
+
+GeoMesa Spark provides capabilities to run geospatial analysis jobs on
+the distributed, large-scala data processing engine `Apache Spark`_. It provides
```
Comment (mdzimmerman) [inline docs/user/spark/architecture.rst]: Heh, muscle memory, I guess. Fixed.

File: docs/user/spark/core.rst (hunk 1 above)
Comment (mdzimmerman) [inline docs/user/spark/core.rst]: Actually, that's just wrong grammar, so no "[picky]" needed! Good catch.

File: docs/user/spark/core.rst (hunk 2 above)
Comment (mdzimmerman) [inline docs/user/spark/core.rst]: Fixed.

File: docs/user/spark/sparksql_functions.rst (hunk 3 above)
Comment (mdzimmerman) [inline docs/user/spark/sparksql_functions.rst]: Yep! Fixed.

File: docs/user/spark/architecture.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
+Architecture
+------------
+
+GeoMesa Spark provides capabilities to run geospatial analysis jobs on
+the distributed, large-scale data processing engine `Apache Spark`_. It provides
```
Comment (jnh5y) [inline docs/user/spark/architecture.rst]: The 'it' here is unclear.  Is 'it' Spark or GeoMesa?

File: docs/user/spark/core.rst [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+`GeoTools ContentDataStore tutorial`_ takes a single parameter called "file". To use
+this data store with GeoMesa Spark, do the following:
+
+.. code-block:: scala
+
+    val params = Map(
+      "geotools" -> "true",
+      "file" -> "locations.csv")
+    val query = new Query("locations")
+    val rdd = GeoMesaSpark(params).rdd(new Configuration(), sc, params, query)
```
Comment (jnh5y) [inline docs/user/spark/core.rst]: Sweet!  Did you try out the CSV DataStore in a REPL?

File: docs/user/spark/core.rst [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+The name of the feature type to access in the data store is passed as the type name of the
+query passed to the ``rdd()`` method. In the example of the `CSVDataStore`_, this is the
+basename of the filename passed as an argument.
+
+.. warning::
+
+    Do not use the GeoTools RDD provider with a GeoMesa Accumulo data store. The
+    :ref:`accumulo_rdd_provider` provides additional optimizations to improve performance
+    between Spark/SparkSQL and GeoMesa Accumulo data stores.
```
Comment (jnh5y) [inline docs/user/spark/core.rst]: NB: (Maybe worth noting in the docs...)  The k-v, "geotools" -> "true" is required to kick over to the GT RDD Provider.

File: docs/user/spark/sparksql.rst [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    // Create DataFrame using the "geomesa" format
+    val dataFrame = sparkSession.read
+      .format("geomesa")
+      .options(dsParams)
+      .option("geomesa.feature", "chicago")
+      .load()
+    dataFrame.createOrReplaceTempView("chicago")
+
+    // Query against the "chicago" schema
+    val sqlQuery = "select * from chicago where st_contains(st_makeBBOX(0.0, 0.0, 90.0, 90.0), geom)"
```
Comment (jnh5y) [inline docs/user/spark/sparksql.rst]: Notable:  bbox does not cover Chicago, IL.

File: docs/user/spark/sparksql.rst [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    */
+
+Usage
+^^^^^
+
+Because GeoMesa SparkSQL stacks on top of the ``geomesa-spark-core`` module,
+one or more of the ``SpatialRDDProvider`` implementations, as described in
+:doc:`/user/spark/core`, must be included on the classpath.
+
+As well, Spark must be configured to register ``GeoMesaSparkKryoRegistor``
```
Comment (jnh5y) [inline docs/user/spark/sparksql.rst]: I think it is more confusing than this...  We need the GeoMesaSparkKryoRegistor if and only if we are passing around RDDs of SimpleFeatures.

In a strict dataframe use case, I *think* we can side step that.

File: docs/user/spark/architecture.rst (hunk 6 above)
Comment (aheyne) [inline docs/user/spark/architecture.rst]: Hmm, good point, I just took 'GeoMesa Spark' as a single entity.

File: docs/user/spark/sparksql_functions.rst [hunk 11]
```diff
@@ -0,0 +1,810 @@
...
+
+
+.. _st_Touches:
```
Comment (jnh5y) [inline docs/user/spark/sparksql_functions.rst]: did we get the case wrong with st_touches?

File: docs/user/spark/sparksql_functions.rst [hunk 12]
```diff
@@ -0,0 +1,810 @@
...
+st_Touches
+^^^^^^^^^^
+
+::
+
+    Boolean st_Touches(Geometry a, Geometry b)
+
+Returns true if the geometries have at least one point in common, but their interiors do not intersect.
+
+.. _st_Within:
```
Comment (jnh5y) [inline docs/user/spark/sparksql_functions.rst]: Ditto for st_Within?

File: docs/user/spark/core.rst (hunk 7 above)
Comment (mdzimmerman) [inline docs/user/spark/core.rst]: No, I didn't, though I should!

File: docs/user/spark/sparksql_functions.rst (hunk 11 above)
Comment (mdzimmerman) [inline docs/user/spark/sparksql_functions.rst]: Yeah, that's a Google Drive export fail.

File: docs/user/spark/sparksql_functions.rst (hunk 12 above)
Comment (mdzimmerman) [inline docs/user/spark/sparksql_functions.rst]: Ditto.

File: docs/user/spark/sparksql.rst (hunk 10 above)
Comment (mdzimmerman) [inline docs/user/spark/sparksql.rst]: Hrmmmm. What if I changed it to *should*?

File: docs/user/spark/architecture.rst (hunk 6 above)
Comment (mdzimmerman) [inline docs/user/spark/architecture.rst]: Yeah, I meant GeoMesa (or the GeoMesa Spark libraries, specifically), and I didn't want to start three sentences in a row with "GeoMesa Spark is ...". I've changed it to something less ambiguously worded.

File: docs/user/spark/core.rst (hunk 8 above)
Comment (mdzimmerman) [inline docs/user/spark/core.rst]: yeah, I clarified a little bit.

File: docs/user/spark/sparksql.rst (hunk 9 above)
Comment (mdzimmerman) [inline docs/user/spark/sparksql.rst]: @cne1x 's negated longitude from the AQS strikes again!

**Pull request merged**